File: Data Structures & Algorithms/binary-tree-diameter/submission-2.go
/**
 * Definition for a binary tree node.
 * type TreeNode struct {
 *     Val int
 *     Left *TreeNode
 *     Right *TreeNode
 * }
 */

func rec(root *TreeNode, ans *int) int{
	if root == nil{
		return 0
	}
	left := rec(root.Left, ans)
	right := rec(root.Right, ans)
	*ans = max(*ans, left + right)
	return max(left, right) + 1
}
func diameterOfBinaryTree(root *TreeNode) int {
	ans := 0
	rec(root, &ans)
	return ans
}

File: Data Structures & Algorithms/binary-tree-maximum-path-sum/submission-2.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right

class Solution:
    def maxPathSum(self, root: Optional[TreeNode]) -> int:
        if not root:
            return 0

        self.ans = float("-inf")
        
        def rec(root):
            if not root:
                return float("-inf")

            left = rec(root.left)
            right = rec(root.right)
            cur = root.val

            a = cur+left+right
            b = cur+left
            c = cur+right
            d = cur
            self.ans = max(self.ans, max(a,b,c,d))
            return max(b,c,d)

        rec(root)
        return self.ans



    
        